File: 14 Longest Common Prefix/14longest-common-prefix.ts
function longestCommonPrefix(strs: string[]): string {
    
    if(strs.length === 1){
        return strs[0]
    }
    
    let ref = strs[0];
    let refIndex = 0;
    
    for(let i = 1; i < strs.length; i++){
        if(strs[i].length < ref.length){
            ref = strs[i]
            refIndex = i
        }
    }
    
    let prefix = "";
    
    
    for(let i = 0; i < strs.length; i++){
        if(refIndex === i){
            continue
        }
        
        for(let j = 0; j < ref.length; j++){            
            if(strs[i][j] === ref[j] && prefix.length === j){
                prefix += ref[j]
            }
            
            else if(strs[i][j] !== ref[j] && j === 0){
                return prefix = ""
            }
            
            else if(strs[i][j] !== ref[j]){
                prefix = prefix.substring(0, j)
                ref = ref.substring(0, j)
                break
            }
        }
    }
    
    return prefix
};

File: 14 Longest Common Prefix/14longest-common-prefix.java
class Solution {
    public String longestCommonPrefix(String[] strs) {
        Arrays.sort(strs);
        
        StringBuilder prefix = new StringBuilder();
        
        String firstString = strs[0];
        String lastString = strs[strs.length-1];
            
        int length1 = firstString.length();
        int length2 = lastString.length();
        
        int min = Math.min(length1, length2);
        
        for(int i = 0; i < min; i++){
            if(firstString.charAt(i) == lastString.charAt(i)){
                prefix.append(firstString.charAt(i));
            } else {
                return prefix.toString();
            }
        }
        
        return prefix.toString();
        
    }
}